File: Beecrowd_Parte4/Permutacao_Ordenadas_1401.java
import java.util.*;

public class Permutacao_Ordenadas_1401 {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        while (n-- > 0) {
            char[] s = sc.next().toCharArray();
            Arrays.sort(s);
            do {
                System.out.println(new String(s));
            } while (nextPermutation(s));
            System.out.println();
        }
        sc.close();
    }

    static boolean nextPermutation(char[] array) {
        for (int i = array.length - 2; i >= 0; i--) {
            if (array[i] < array[i + 1]) {
                for (int j = array.length - 1; ; j--) {
                    if (array[j] > array[i]) {
                        swap(array, i, j);
                        reverse(array, i + 1, array.length - 1);
                        return true;
                    }
                }
            }
        }
        return false;
    }

    static void swap(char[] array, int i, int j) {
        char temp = array[i];
        array[i] = array[j];
        array[j] = temp;
    }

    static void reverse(char[] array, int i, int j) {
        while (i < j) {
            swap(array, i++, j--);
        }
    }
}

File: Beecrowd_Parte4/Permutacao_Ordenadas_1401.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int compare(const void *a, const void *b) {
    return *(char *)a - *(char *)b;
}

void swap(char *a, char *b) {
    char temp = *a;
    *a = *b;
    *b = temp;
}

int nextPermutation(char *str, int size) {
    int i = size - 2;
    while (i >= 0 && str[i] >= str[i + 1]) {
        i--;
    }
    if (i < 0) {
        return 0;
    }
    int j = size - 1;
    while (str[j] <= str[i]) {
        j--;
    }
    swap(&str[i], &str[j]);
    for (int a = i + 1, b = size - 1; a < b; a++, b--) {
        swap(&str[a], &str[b]);
    }
    return 1;
}

int main() {
    int n;
    scanf("%d", &n);
    while (n--) {
        char str[11];
        scanf("%s", str);
        int size = strlen(str);
        qsort(str, size, sizeof(char), compare);
        do {
            printf("%s\n", str);
        } while (nextPermutation(str, size));
        printf("\n");
    }
    return 0;
}